controller/system_rollout_controller.go

```diff
@@ -0,0 +1,1788 @@
+	}
+
+	if c.engineImage != "" {
+		log = log.WithField("engineImage", c.engineImage)
+	}
+
+	return log
+}
+
+func (c *SystemRolloutController) recordErrorState(record *systemRolloutRecord, systemRestore *longhorn.SystemRestore, err string, log logrus.FieldLogger) {
+	systemRestore.Status.State = longhorn.SystemRestoreStateError
+	systemRestore.Status.Conditions = types.SetCondition(
+		systemRestore.Status.Conditions,
+		longhorn.SystemRestoreConditionTypeError,
+		longhorn.ConditionStatusTrue,
+		record.reason,
+		fmt.Sprintf("%v: %v", record.message, err),
+	)
+
+	log.WithError(fmt.Errorf(err)).Error(record.message)
+	c.eventRecorder.Eventf(systemRestore, corev1.EventTypeWarning, EventReasonFailed, util.CapitalizeFirstLetter(record.message))
+
+	c.cacheErrors.Reset()
+}
+
+func (c *SystemRolloutController) recordNormalState(record *systemRolloutRecord, systemRestore *longhorn.SystemRestore, err string, log logrus.FieldLogger) {
+	systemRestore.Status.State = record.nextState
+	log.Info(record.message)
+	c.eventRecorder.Eventf(systemRestore, corev1.EventTypeNormal, record.reason, record.message)
+}
+
+func (c *SystemRolloutController) updateSystemRolloutRecord(record *systemRolloutRecord, recordType systemRolloutRecordType, nextState longhorn.SystemRestoreState, reason, message string) {
+	record.recordType = recordType
+	record.nextState = nextState
+	record.reason = reason
+	record.message = message
+}
+
+func (c *SystemRolloutController) systemRollout() error {
+	log := c.getLoggerForSystemRollout()
+
+	systemRestore, err := c.ds.GetSystemRestore(c.systemRestoreName)
+	if err != nil {
+		return err
+	}
+
+	record := &systemRolloutRecord{}
+	existingSystemRestore := systemRestore.DeepCopy()
+	defer func() {
+		c.handleStatusUpdate(record, systemRestore, existingSystemRestore, err, log)
+	}()
+
+	switch systemRestore.Status.State {
+	case longhorn.SystemRestoreStatePending:
+		systemRestore, err = c.initializeSystemRollout(systemRestore, log)
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize system rollout")
+		}
+
+		c.updateSystemRolloutRecord(record,
+			systemRolloutRecordTypeNormal, longhorn.SystemRestoreStateDownloading,
+			EventReasonStart, fmt.Sprintf(SystemRolloutMsgInitializedFmt, systemRestore.Spec.SystemBackup),
+		)
+
+	case longhorn.SystemRestoreStateDownloading:
+		err = c.Download(log)
+		if err != nil {
+			return errors.Wrap(err, "failed to download system backup")
+		}
+
+		c.updateSystemRolloutRecord(record,
+			systemRolloutRecordTypeNormal, longhorn.SystemRestoreStateUnpacking,
+			EventReasonFetched, fmt.Sprintf(SystemRolloutMsgDownloadedFmt, c.backupTargetClient.GetURL()),
+		)
+
+	case longhorn.SystemRestoreStateUnpacking:
+		err = c.Unpack(log)
+		if err != nil {
+			c.updateSystemRolloutRecord(record,
+				systemRolloutRecordTypeError, longhorn.SystemRestoreStateError,
+				longhorn.SystemRestoreConditionReasonUnpack, longhorn.SystemRestoreConditionMessageUnpackFailed,
+			)
+			return nil
+		}
```

Comment:
Not sure if I understand the question, can you elaborate a bit more?

Reply:
For example, in case `longhorn.SystemRestoreStateUnpacking`, the function will execute `c.updateSystemRolloutRecord` when there is an error (line 438), but won't in case `longhorn.SystemRestoreStateDownloading` (line 427)

Reply:
Where I've made to return the error is where I think the situation is recoverable by `handleErr` so we don't need to mark it in Error state.

Reply:
Got it. Please have a simple comment for it.